### PR TITLE
Remove `Basic.compare` from physics and printing

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

This is more redux from #19907 to only handle the issues with physics and printing.
I just nullified the keyword 'old' from printing

Also some ordering in physics.physics are changed I assume, 
but I think that it is better than the changes in #19907 about physics because I flipped the sorting such that it correctly sorts things alphabetical at least
for example, `Commutator(A, B)` is canonical and it isn't wrongly written as `-Commutator(B, A)`

#### Other comments

I also recognize that `order=None` and `order='none'` are different and confusing options for printing, maybe this is abused and can cause some buggy behavior, so I'm not sure 'old' have to be removed and it can replace 'none' instead.

I only see that `old` exists to do some more redundant computation
https://github.com/sympy/sympy/blob/1d2de06782fe1f46f1fb9b211b406431f666d221/sympy/printing/printer.py#L345-L346
so removing this wont be difficult
so I'll leave 'old' keyword should be something like we 'dont care' about how it is printed in the future.

#### Release Notes

- interactive
  - BREAKING: The legacy option `order='old'` is deprecated in `init_printing, init_session`. Use `order='none'` instead.
- printing
  - BREAKING: The legacy option `order='old'` is deprecated in all printers. Use `order='none'` instead.

### CODE REVIEW

File: sympy/printing/latex.py [hunk 1]
```diff
@@ -2978,11 +2978,9 @@ def latex(expr, **settings):
         ``'ldot'``, ``'dot'``, or ``'times'``.
     order: string, optional
         Any of the supported monomial orderings (currently ``'lex'``,
-        ``'grlex'``, or ``'grevlex'``), ``'old'``, and ``'none'``. This
-        parameter does nothing for `~.Mul` objects. Setting order to ``'old'``
-        uses the compatibility ordering for ``~.Add`` defined in Printer. For
-        very large expressions, set the ``order`` keyword to ``'none'`` if
-        speed is a concern.
+        ``'grlex'``, or ``'grevlex'``), and ``'none'``. This parameter does
+        nothing for `~.Mul` objects. For ery large expressions, set the
```
Comment (oscarbenjamin) [inline sympy/printing/latex.py:2982]: very

**Comment by asmeurer** [comment]
`order='old'` is literally the "older" ordering that was used in SymPy before the option to do different ordering was added. Changing the semantics of 'old' would defeat the purpose. If we can't support it anymore we should just remove it. Anyway it's so old at this point that I don't know if anyone uses it. The main benefit of it is it prints polynomials in reverse lex order, but you can and should do that with "rev-lex". A big problem is that `order` is very poorly documented so I don't know how many people know about `rev-lex`. But maybe `order='old'` isn't so well documented either.

There's no relationship to `order='none'`, which just disables sorting entirely. `order=None` just means the default option (it is an unfortunate name).

**Comment by oscarbenjamin** [comment]
> There's no relationship to `order='none'`, which just disables sorting entirely. `order=None` just means the default option (it is an unfortunate name).

We should come up with a better name for this ordering and document it prominently. Every time I see a user wanting a different ordering none of the monomial orderings is of any interest to them. They usually just want the expression to print as it is (i.e. `order='none'`). Sometimes they would like better handling of things like `-x + y`. I never see anyone who is interested in grevlex etc. Those are used by Poly which is fine but there is no need for those in general expression printing. Personally I would default to `order='none'` not just for performance reasons but more importantly because I would rather see something that resembles the actual representation of the expression.

**Comment by sylee957** [comment]
Apparently, some people made recent patches that abuse `order='old'` 
https://github.com/sympy/sympy/blob/b1de2706ce71dfdb332a7286110aca518c29c2d7/sympy/printing/tests/test_latex.py#L207-L215

**Comment by sylee957** [comment]
> Anyway it's so old at this point that I don't know if anyone uses it.

I'm fine with raising errors though, but apparently that is an example that that can be some assumption.

**Comment by asmeurer** [comment]
> Personally I would default to order='none' not just for performance reasons but more importantly because I would rather see something that resembles the actual representation of the expression.

I'd like to also somehow move to Add and Mul not doing any sorting in their constructors. I think it can be done by storing the args as a frozenset and internally only using that frozenset, and only sorting it into a tuple `args` when requested for backwards compatibility. And and Or already more or less do this, so it can work. 

For the printers, I think the only ordering that people really care about is very basic stuff, namely:

- Polynomial terms should be printed in lexicographic order
- Terms with minus signs should print after terms without them (assuming it doesn't conflict with the previous point)

Both of these can be done very cheaply. The problem is that the current design tries to impose a global order on every possible SymPy expression. That sort of thing can be useful in some cases, but it's really not that important for printing. It really doesn't matter if `expr1` gets printed before `expr2` in one expression and after it in another expression. 

So I think we need a new default printing order, that just tries to do these very basic things, and only in a way that is very cheap (and definitely does *not* do any numerical evaluation). If a user really wants a consistent global ordering, they can use one of the other order arguments or a custom sort key.

**Comment by oscarbenjamin** [comment]
> I think the only ordering that people really care about is very basic stuff, namely:
> 
> * Polynomial terms should be printed in lexicographic order

I've never seen anyone care about this.

> I'd like to also somehow move to Add and Mul not doing any sorting in their constructors. I think it can be done by storing the args as a frozenset and internally only using that frozenset, and only sorting it into a tuple `args` when requested for backwards compatibility

No, it should be possible to specify the order of the arguments. This is the behaviour that users (those who care) want to have:
```python
In [1]: init_printing(order='none')

In [2]: Add(y, x, evaluate=False)
Out[2]: y + x
```
They actually want to have that behaviour without calling `init_printing` or using `evaluate=False`.

Yes, some would prefer to have more careful handling of minus signs but of those who would use the `order` parameter for the printers 99% are looking for the behaviour above. They want Add to preserve the ordering of the args and they want the printer to show them in the same order.

**Comment by sylee957** [comment]
The problem is, it is not much desirable to recommend users to use ambiguous name like `'none'` to fill in that gap, because that could also likely be deprecated in the future because of confusion.
I think that it makes things to grow out of scale how to find a better way to substitute such things.
I suggested the original patch just to silently fill in `'old' == 'none'`, than raising errors. at the point now, if it doesn't get oppositions.

**Comment by oscarbenjamin** [comment]
> The problem is, it is not much desirable to recommend users to use ambiguous name like `'none'` instead, because that could also likely be deprecated in the future because of confusion.

I would suggest something like `order = 'raw'`, `order = 'unmodified'` or `order = 'plain'` but it should have the same meaning that `order = 'none'` currently has.

**Comment by sylee957** [comment]
Unfortunately, after browsing the code, the change from `'none' == 'raw'` looks like very large change because it is also used in other libraries like `cse` so changing this stuff also breaks other consistency make the work unnecessarily grow bigger.
I'd just leave the PR now to wait for any approval.

**Comment by oscarbenjamin** [comment]
Both `'none'` and `'raw'` can continue to be allowed as different names for the same thing. The point is to offer a name as a suggestion in the documentation.

**Comment by asmeurer** [comment]
> I've never seen anyone care about this.

That's because it already does it. If we removed it, you'd see people complaining about it. The 'old' order, which as I mentioned used to be the default, prints things in reverse lexicographic order (this was chosen because of series expansions, which was the first primary use-case for SymPy). There were plenty of people who complained that polynomials printed backwards (myself included), and then it was fixed. 

> No, it should be possible to specify the order of the arguments. This is the behaviour that users (those who care) want to have:

Ideally we could do both. You really want to compare sets when doing equality checking. That's much faster than sorting. Unfortunately, unlike dicts, Python sets do not maintain insertion order, so this would require some internal hacks (but it would be doable).

Unless you are also suggesting that `x + y == y + x` give `False`. That would be so breaking I'm very sceptical we could make it work. 

> I suggested the original patch just to silently fill in 'old' == 'none', than raising errors. at the point now, if it doesn't get oppositions.

That already breaks compatibility, but in a way that doesn't make sense. `'none'` was never the "old" way of sorting. Even if it did make sense we really should stop naming APIs "new" and "old". It's undescriptive, and becomes invalid as time goes on.

If we are going to break compatibility with `'old'`, we should just do it by removing it entirely. But since this won't happen before the release, we might as well deprecate it first. 

The alternative is to keep 'old' around unchanged, but I don't see a value to this since keeping it working after we remove BasicMeta would just mean writing a shim to emulate it, and any benefit of 'old' can be gotten by a more explicit polynomial ordering or custom sort key.

**Comment by oscarbenjamin** [comment]
> > I've never seen anyone care about this.
> 
> That's because it already does it. If we removed it, you'd see people complaining about it

*None* of the monomial orderings makes sense if you don't already have an ordering on symbols:
```python
In [25]: a, b, c, d = symbols('a, b, c, d')

In [26]: l = symbols('lambda')

In [27]: Matrix([[a, b], [c, d]]).charpoly(l).as_expr()
Out[27]: 
             2             
a⋅d - b⋅c + λ  + λ⋅(-a - d)
```
I imagine that people used to complain for precisely the reason that they do now: sometimes they want to control the order but SymPy (both through canonicalisation and printing) makes that impossible or at least difficult.

**Comment by asmeurer** [comment]
This would put the burden on any library function that creates an expression to put it in a nice order. Right now that logic is centralized in the printers. I do agree it would be nice if we could make things like characteristic polynomials actually come out in the right order. I do see a lot of people requesting to fudge the printing order in various ways, but maybe those people would all be satisfied if they could just fix it by creating the expression itself in the order they want.

By the way, if the plan is to make `order='none'` the new default, then there's no need for a new name. `'none'` can just become literally `None`.

**Comment by sylee957** [comment]
@asmeurer I just open other PRs #24606 and #24607 separately to patch the library code to stay away from BasicMeta.

**Comment by sylee957** [comment]
This is little bit redundant now because `BasicMeta` was removed in `master` in alternative way.
I think that replacing `order='old'` can be done in better dedicated pull request.
For the `physics/quantum` stuff, I think that `default_sort_key` can be better drop-in replacement than `Basic.compare`, but I'd only touch this if some limitation occurs for that.

**Pull request closed without merging**